File: server/src/models/SpotlightAudiobook.js
/**
 * Spotlight Audiobook Model
 * Featured audiobooks that can be promoted without app rebuilds
 */
export class SpotlightAudiobook {
  constructor(data) {
    this.id = data.id;
    this.audiobookSnippetId = data.audiobookSnippetId; // Reference to the audiobook snippet
    this.title = data.title || '';
    this.priority = data.priority || 0; // Higher priority = shown first
    this.isActive = data.isActive ?? true;
    this.startDate = data.startDate ? new Date(data.startDate) : new Date();
    this.endDate = data.endDate ? new Date(data.endDate) : null;
    this.promotionText = data.promotionText || ''; // Custom promotion message
    this.badge = data.badge || null; // e.g., "New Release", "Editor's Pick"
    this.targetAudience = data.targetAudience || 'all';
    this.createdAt = data.createdAt || new Date();
    this.updatedAt = data.updatedAt || new Date();
  }

  toFirestore() {
    return {
      audiobookSnippetId: this.audiobookSnippetId,
      title: this.title,
      priority: this.priority,
      isActive: this.isActive,
      startDate: this.startDate,
      endDate: this.endDate,
      promotionText: this.promotionText,
      badge: this.badge,
      targetAudience: this.targetAudience,
      createdAt: this.createdAt,
      updatedAt: this.updatedAt,
    };
  }

  static fromFirestore(doc) {
    const data = doc.data();
    return new SpotlightAudiobook({
      id: doc.id,
      ...data,
    });
  }

  isCurrentlyActive() {
    if (!this.isActive) return false;
    const now = new Date();
    if (now < this.startDate) return false;
    if (this.endDate && now > this.endDate) return false;
    return true;
  }
}

File: lib/models/spotlight_audiobook.dart
import 'audiobook_snippet.dart';

class SpotlightAudiobook {
  final String id;
  final String audiobookSnippetId;
  final String title;
  final int priority;
  final bool isActive;
  final DateTime startDate;
  final DateTime? endDate;
  final String promotionText;
  final String? badge;
  final String targetAudience;
  final DateTime createdAt;
  final DateTime updatedAt;
  final AudiobookSnippet? snippet;
  
  SpotlightAudiobook({
    required this.id,
    required this.audiobookSnippetId,
    required this.title,
    required this.priority,
    required this.isActive,
    required this.startDate,
    this.endDate,
    required this.promotionText,
    this.badge,
    required this.targetAudience,
    required this.createdAt,
    required this.updatedAt,
    this.snippet,
  });
  
  factory SpotlightAudiobook.fromJson(Map<String, dynamic> json) {
    return SpotlightAudiobook(
      id: json['id'] as String,
      audiobookSnippetId: json['audiobookSnippetId'] as String,
      title: json['title'] as String? ?? '',
      priority: json['priority'] as int? ?? 0,
      isActive: json['isActive'] as bool? ?? true,
      startDate: json['startDate'] != null
          ? DateTime.parse(json['startDate'].toString())
          : DateTime.now(),
      endDate: json['endDate'] != null
          ? DateTime.parse(json['endDate'].toString())
          : null,
      promotionText: json['promotionText'] as String? ?? '',
      badge: json['badge'] as String?,
      targetAudience: json['targetAudience'] as String? ?? 'all',
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'].toString())
          : DateTime.now(),
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'].toString())
          : DateTime.now(),
      snippet: json['snippet'] != null
          ? AudiobookSnippet.fromJson(json['snippet'] as Map<String, dynamic>)
          : null,
    );
  }
  
  bool isCurrentlyActive() {
    if (!isActive) return false;
    final now = DateTime.now();
    if (now.isBefore(startDate)) return false;
    if (endDate != null && now.isAfter(endDate!)) return false;
    return true;
  }
}

